src/transport/types.ts

```diff
@@ -0,0 +1,73 @@
+import events from 'events'
+import Multiaddr from 'multiaddr'
+import Connection from '../connection/connection'
+import { Sink } from '../stream-muxer/types'
+
+export type DialOptions = {
+  signal?: AbortSignal
+}
+
+/**
+ * A libp2p transport is understood as something that offers a dial and listen interface to establish connections.
+ */
+export interface Transport <DialOptions extends { signal?: AbortSignal }> {
+  new (upgrader: Upgrader, ...others: any): Transport<DialOptions>; // eslint-disable-line
+  prototype: Transport <DialOptions>;
```

Comment:
Is this really necessary ? If not I'd get rid of it

Reply:
Yes, here we don't have any implementation just the interface, so if a method receives Transport reference and create an instance the type needs to be constructable

Reply:
FYI I am trying out @Gozala's suggestion with the factory below. It seems a better solution for this